docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+
+Preconditions:
+- `Spec.Endpoint` is empty
+- `Spec.LoadBalancerRef` is nil
+
+Reconciliation:
+TODO
+
+### New Types
+
+#### Load Balancer
+
+A common type to aggregate information between Load Balancer provider implementations and to provide some common reconciliation logic.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: LoadBalancer
+  metadata:
+    name: my-cluster-control-plane-lb
+spec:
+  clusterName: my-cluster
+  endpoint:
+    host: test.example.com
+    port: 443
+  targetKind:
+    apiGroup: cluster.x-k8s.io
+    kind: Machine
+  selector:
+    matchExpressions:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      loadbalancer.cluster.x-k8s.io/name: my-cluster-control-plane-lb
+  infrastructureRef:
+    apiGroup: infrastructure.cluster.x-k8s.io
+    kind: AWSNLB
+    name: my-cluster-control-plane-nlb
+status:
+  targetKind: <string representing group/kind name for selector>
+  targetSelector: <string representing the label selector for the referenced LB>
+  targets: (List of members matching label selector)
+  activeTargets: (List of active members)
+  conditions: (List of infra providers condition for the load balancer)
+```
+
+#### Annotations
+
+- Used by providers to signal the state of the resource in the LB
+  - loadbalancer.cluster.x-k8s.io/name: `<name of lb>`
+  - loadbalancer.cluster.x-k8s.io/status: `<status>`
+    - Statuses:
+      - 'active' (configured and active in configuration)
+      - 'removed' (removed from configuration and safe to delete)
+      - 'disabled' (user requested disable complete)
+      - other provider-relevant statuses for in-between states
+- Used by users/external systems to signal to the load balancer provider that a resource should be temporarily removed from the LB:
+  - loadbalancer.cluster.x-k8s.io/disabled:
+
+#### Conditions
+
+TODO
+
+#### Defaults
+
+- If Spec.Selector is not defined:
+  - Spec.Selector.MatchLabels = `{'cluster.x-k8s.io/cluster-name': Spec.ClusterName, 'loadbalancer.cluster.x-k8s.io/name': Metadata.Name }`
+
+#### Watches:
+
+- Cluster
+- LoadBalancer
+- ??? Kind referenced by LB selector
+- LoadBalancer provider types
+
+#### Reconciliation
+
+![Load Balancer Reconciliation Flow](images/load-balancer/lb-reconciliation.png)
+
+###### Scenario: externally managed endpoint
+
+###### Scenario: provider managed endpoint
+
+
+TODO: gate provider load balancer reconciliation when user defined endpoint hasn't yet been pushed down to LB Provider
+
+### Example Load Balancer Provider (FooLoadBalancer)
+In this example, we are provisioning a Kubernetes cluster on AWS to be backed by a fully hypothetical Foo Load Balancer.
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: Cluster
+metadata:
+  name: my-cluster
+  namespace: default
+spec:
+  clusterNetwork:
+    pods:
+      cidrBlocks:
+        - 192.168.0.0/16
+  controlPlaneRef:
```

Comment:
It seems the examples here do not align with the proposes changes to cluster here https://github.com/kubernetes-sigs/cluster-api/pull/4389/files#diff-3dabbc43888d2653843c8851f0293425b1edcfa8d613d654d5a8f1dbab17e979R173